File: routeMap/app/src/main/java/ru/android/zheka/coreUI/AbstractFragment.java
package ru.android.zheka.coreUI;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

public abstract class AbstractFragment<B extends ViewDataBinding> extends Fragment implements IBaseFragment {
    private ErrorControl error;
    private B binding;

    protected abstract int getLayoutId();

    protected abstract void initComponent();

    protected abstract void onInitBinding(B binding);

    protected abstract void onResumeBinding(B binding);

    protected abstract void onDestroyBinding(B binding);

    @Override
    public void onStart() {
        super.onStart ();
        initComponent ();
        onInitBinding (binding);
    }

    @Override
    public void onResume() {
        super.onResume ();
        onResumeBinding (binding);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView ();
        onDestroyBinding (binding);
        binding.unbind ();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate (savedInstanceState);
        error = new ErrorControl (this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle saveInstance) {
        binding = DataBindingUtil.inflate (inflater, getLayoutId (), container, false);
        return binding.getRoot ();
    }

    public void showError(Throwable throwable) {
        error.showError (throwable, a -> {
        });
    }

    public Context getContext() {
        return getActivity ();
    }

    @Override
    public void switchToFragment(int fragmentId, @NonNull Fragment fragment) {
        FragmentTransaction transaction = getManager ().beginTransaction ();
        transaction.replace (fragmentId, fragment);
        transaction.commit ();
    }

    @NonNull
    @Override
    public FragmentManager getManager() {
        return getActivity ().getSupportFragmentManager ();
    }

}

File: routeMap/app/src/main/java/ru/android/zheka/coreUI/AbstractFragment.kt
package ru.android.zheka.coreUI

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import dagger.android.support.DaggerFragment
import io.reactivex.functions.Consumer

abstract class AbstractFragment<B : ViewDataBinding> : DaggerFragment(), IActivity {
    private var error: ErrorControl? = null
    lateinit var binding: B
    protected abstract val layoutId: Int
    protected abstract fun initComponent()
    protected abstract fun onInitBinding(binding: B)
    protected abstract fun onResumeBinding(binding: B)
    protected abstract fun onDestroyBinding(binding: B)
    override fun onStart() {
        super.onStart()
        initComponent()
        onInitBinding(binding)
    }

    override fun context() = super.getActivity()!! as Context

    override fun onResume() {
        super.onResume()
        onResumeBinding(binding)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        onDestroyBinding(binding)
        binding.unbind()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        error = ErrorControl(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, saveInstance: Bundle?): View? {
        binding = DataBindingUtil.inflate(inflater, layoutId, container, false)
        binding = initAdapter(binding)
        return binding.getRoot()
    }

    open fun initAdapter(binding: B):B {return binding}

    override fun showError(throwable: Throwable) {
        error!!.showError(throwable, Consumer{ a: Boolean? -> })
    }

    override fun switchToFragment(fragmentId: Int, fragment: Fragment) {
        val transaction = manager!!.beginTransaction()
        transaction.replace(fragmentId, fragment)
        transaction.commit()
    }

    override fun removeFragment(fragment: Fragment) {
        val transaction = manager!!.beginTransaction()
        transaction.remove(fragment)
        transaction.commit()
    }

    override val manager
        get() = fragmentManager

    override val activity
        get() = super.getActivity()?:Activity()
}